adapters/tlsio_ssl_esp8266.c

```diff
@@ -0,0 +1,888 @@
+    ON_IO_ERROR on_io_error;
+    void* on_bytes_received_context;
+    void* on_io_open_complete_context;
+    void* on_io_close_complete_context;
+    void* on_io_error_context;
+    SSL* ssl;
+    SSL_CTX* ssl_context;
+    TLSIO_STATE tlsio_state;
+    char* hostname;
+    int port;
+    char* certificate;
+    const char* x509certificate;
+    const char* x509privatekey;
+    int sock;
+    ip_addr_t target_ip;
+} TLS_IO_INSTANCE;
+
+/*this function destroys an option previously created*/
+static void tlsio_openssl_DestroyOption(const char* name, const void* value)
+{
+    /*since all options for this layer are actually string copies, disposing of one is just calling free*/
+    if (
+        (name == NULL) || (value == NULL)
+        )
+    {
+        LogError("invalid parameter detected: const char* name=%p, const void* value=%p", name, value);
+    }
+    else
+    {
+        if (
+            (strcmp(name, "TrustedCerts") == 0) ||
+            (strcmp(name, "x509certificate") == 0) ||
+            (strcmp(name, "x509privatekey") == 0)
+        )
+        {
+            free((void*)value);
+        }
+        else
+        {
+            LogError("not handled option : %s", name);
+        }
+    }
+}
+
+
+/* Codes_SRS_TLSIO_SSL_ESP8266_99_078: [ The tlsio_openssl_retrieveoptions shall not do anything, and return NULL. ]*/
+static OPTIONHANDLER_HANDLE tlsio_openssl_retrieveoptions(CONCRETE_IO_HANDLE tlsio_handle)
+{
+    (void)(tlsio_handle);
+        
+    /* Not implementing any options */
+    return NULL;
+}
+
+static const IO_INTERFACE_DESCRIPTION tlsio_openssl_interface_description =
+{
+    tlsio_openssl_retrieveoptions,
+    tlsio_openssl_create,
+    tlsio_openssl_destroy,
+    tlsio_openssl_open,
+    tlsio_openssl_close,
+    tlsio_openssl_send,
+    tlsio_openssl_dowork,
+    tlsio_openssl_setoption
+};
+
+static void indicate_open_complete(TLS_IO_INSTANCE* tls_io_instance, IO_OPEN_RESULT open_result)
+{
+    if (tls_io_instance->on_io_open_complete == NULL)
+    {
+        LogError("NULL on_io_open_complete.");
+    }
+    else
+    {
+        /* Codes_TLSIO_SSL_ESP8266_99_002: [ The tlsio_ssl_esp8266 shall report the open operation status using the IO_OPEN_RESULT enumerator defined in the `xio.h`.]*/
+        tls_io_instance->on_io_open_complete(tls_io_instance->on_io_open_complete_context, open_result);
+    }
+}
+
+
+static int lwip_net_errno(int fd)
+{
+    int sock_errno = 0;
+    u32_t optlen = sizeof(sock_errno);
+    getsockopt(fd, SOL_SOCKET, SO_ERROR, &sock_errno, &optlen);
+    return sock_errno;
+}
+
+static void lwip_set_non_block(int fd) 
+{
+  int flags = -1;
+  int error = 0;
+
+  while(1){
+      flags = fcntl(fd, F_GETFL, 0);
+      if (flags == -1){
+          error = lwip_net_errno(fd);
+          if (error != EINTR){
+              break;
+          }
+      } else{
+          break;
+      }
+  }
+
+  while(1){
+      flags = fcntl(fd, F_SETFL, flags | O_NONBLOCK);
```

Comment:
@JetstreamRoySprowl Thanks for the feedback. I have updated this block to only call fcntl F_SETFL once. `fcntl(fd, F_SETFL, O_NONBLOCK);`